File: examples/sudoku/solver.c
#include <stdio.h>
#include <stdbool.h>

typedef struct {
  int board[9][9];
  bool rows[9][9];
  bool cols[9][9];
  bool blocks[3][3][9];
} Solver;

void showCell(int cell) {
  if (cell > 0) {
    printf("%d ", cell);
  } else {
    printf("  ");
  }
}

void showRow(int row[]) {
  for (int c = 0; c < 3; c++) {
    showCell(row[c]);
  }

  printf("| ");

  for (int c = 3; c < 6; c++) {
    showCell(row[c]);
  }

  printf("| ");

  for (int c = 6; c < 9; c++) {
    showCell(row[c]);
  }

  printf("\n");
}

void show(int board[9][9]) {
  for (int r = 0; r < 3; r++) {
    showRow(board[r]);
  }

  printf("------+-------+------\n");

  for (int r = 3; r < 6; r++) {
    showRow(board[r]);
  }

  printf("------+-------+------\n");

  for (int r = 6; r < 9; r++) {
    showRow(board[r]);
  }
}

bool solve(Solver* solver, int r, int c) {
  if (c == 9) {
    c = 0;
    r++;
  }

  if (r == 9) {
    return true;
  }

  if (solver->board[r][c] > 0) {
    return solve(solver, r, c + 1);
  }

  bool* row = solver->rows[r];
  bool* col = solver->cols[c];
  bool* block = solver->blocks[r / 3][c / 3];

  for (int n = 1; n <= 9; n++) {
    int i = n - 1;

    if (!(row[i] || col[i] || block[i])) {
      solver->board[r][c] = n;
      row[i] = true;
      col[i] = true;
      block[i] = true;

      if (solve(solver, r, c + 1)) {
        return true;
      }

      solver->board[r][c] = 0;
      row[i] = false;
      col[i] = false;
      block[i] = false;
    }
  }

  return false;
}

int main() {
  // Solver solver = {
  //   .board = {
  //     { 0, 0, 0, 0, 0, 0, 0, 0, 0 },
  //     { 0, 0, 0, 0, 0, 3, 0, 8, 5 },
  //     { 0, 0, 1, 0, 2, 0, 0, 0, 0 },
  //     { 0, 0, 0, 5, 0, 7, 0, 0, 0 },
  //     { 0, 0, 4, 0, 0, 0, 1, 0, 0 },
  //     { 0, 9, 0, 0, 0, 0, 0, 0, 0 },
  //     { 5, 0, 0, 0, 0, 0, 0, 7, 3 },
  //     { 0, 0, 2, 0, 1, 0, 0, 0, 0 },
  //     { 0, 0, 0, 0, 4, 0, 0, 0, 9 },
  //   }
  // };

  Solver solver = {
    .board = {
      { 5, 3, 0, 0, 7, 0, 0, 0, 0 },
      { 6, 0, 0, 1, 9, 5, 0, 0, 0 },
      { 0, 9, 8, 0, 0, 0, 0, 6, 0 },
      { 8, 0, 0, 0, 6, 0, 0, 0, 3 },
      { 4, 0, 0, 8, 0, 3, 0, 0, 1 },
      { 7, 0, 0, 0, 2, 0, 0, 0, 6 },
      { 0, 6, 0, 0, 0, 0, 2, 8, 0 },
      { 0, 0, 0, 4, 1, 9, 0, 0, 5 },
      { 0, 0, 0, 0, 8, 0, 0, 7, 9 },
    }
  };

  for (int r = 0; r < 9; r++) {
    for (int c = 0; c < 9; c++) {
      int i = solver.board[r][c] - 1;

      if (i >= 0) {
        solver.rows[r][i] = true;
        solver.cols[c][i] = true;
        solver.blocks[r / 3][c / 3][i] = true;
      }
    }
  }

  solve(&solver, 0, 0);
  show(solver.board);
  return 0;
}

File: examples/sudoku/solver.rs
#[derive(Default)]
struct Solver {
    board: [[u8; 9]; 9],
    rows: [[bool; 9]; 9],
    cols: [[bool; 9]; 9],
    blocks: [[[bool; 9]; 3]; 3],
}

fn show_cell(cell: u8) {
    if cell > 0 {
        print!("{cell} ");
    } else {
        print!("  ")
    }
}

fn show_row(row: &[u8; 9]) {
    for &cell in &row[0..3] {
        show_cell(cell);
    }

    print!("| ");

    for &cell in &row[3..6] {
        show_cell(cell);
    }

    print!("| ");

    for &cell in &row[6..9] {
        show_cell(cell);
    }

    println!();
}

fn show(board: &[[u8; 9]; 9]) {
    for row in &board[0..3] {
        show_row(row);
    }

    println!("------+-------+------");

    for row in &board[3..6] {
        show_row(row);
    }

    println!("------+-------+------");

    for row in &board[6..9] {
        show_row(row);
    }
}

fn solve(solver: &mut Solver, mut r: usize, mut c: usize) -> bool {
    if c == 9 {
        c = 0;
        r += 1;
    }

    if r == 9 {
        return true;
    }

    if solver.board[r][c] > 0 {
        return solve(solver, r, c + 1);
    }

    for n in 1..=9_u8 {
        let i = (n - 1) as usize;
        let conflict = solver.rows[r][i] || solver.cols[c][i] || solver.blocks[r / 3][c / 3][i];

        if !conflict {
            solver.board[r][c] = n;
            solver.rows[r][i] = true;
            solver.cols[c][i] = true;
            solver.blocks[r / 3][c / 3][i] = true;

            if solve(solver, r, c + 1) {
                return true;
            }

            solver.board[r][c] = 0;
            solver.rows[r][i] = false;
            solver.cols[c][i] = false;
            solver.blocks[r / 3][c / 3][i] = false;
        }
    }

    false
}

fn main() {
    let board = [
        [0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 3, 0, 8, 5],
        [0, 0, 1, 0, 2, 0, 0, 0, 0],
        [0, 0, 0, 5, 0, 7, 0, 0, 0],
        [0, 0, 4, 0, 0, 0, 1, 0, 0],
        [0, 9, 0, 0, 0, 0, 0, 0, 0],
        [5, 0, 0, 0, 0, 0, 0, 7, 3],
        [0, 0, 2, 0, 1, 0, 0, 0, 0],
        [0, 0, 0, 0, 4, 0, 0, 0, 9],
    ];

    // let board = [
    //     [5, 3, 0, 0, 7, 0, 0, 0, 0],
    //     [6, 0, 0, 1, 9, 5, 0, 0, 0],
    //     [0, 9, 8, 0, 0, 0, 0, 6, 0],
    //     [8, 0, 0, 0, 6, 0, 0, 0, 3],
    //     [4, 0, 0, 8, 0, 3, 0, 0, 1],
    //     [7, 0, 0, 0, 2, 0, 0, 0, 6],
    //     [0, 6, 0, 0, 0, 0, 2, 8, 0],
    //     [0, 0, 0, 4, 1, 9, 0, 0, 5],
    //     [0, 0, 0, 0, 8, 0, 0, 7, 9],
    // ];

    let mut solver = Solver {
        board,
        ..Solver::default()
    };

    for r in 0..9 {
        for c in 0..9 {
            if board[r][c] > 0 {
                let i = (board[r][c] - 1) as usize;
                solver.rows[r][i] = true;
                solver.cols[c][i] = true;
                solver.blocks[r / 3][c / 3][i] = true;
            }
        }
    }

    solve(&mut solver, 0, 0);
    show(&solver.board);
}
